source/common/network/apple_dns_impl.cc

```diff
@@ -209,153 +105,65 @@ ActiveDnsQuery* AppleDnsResolverImpl::resolve(const std::string& dns_name,
   return nullptr;
 }
 
-void AppleDnsResolverImpl::addPendingQuery(PendingResolution* query) {
-  ASSERT(queries_with_pending_cb_.count(query) == 0);
-  queries_with_pending_cb_.insert(query);
-}
-
-void AppleDnsResolverImpl::removePendingQuery(PendingResolution* query) {
-  auto erased = queries_with_pending_cb_.erase(query);
-  ASSERT(erased == 1);
-}
-
-void AppleDnsResolverImpl::flushPendingQueries(const bool with_error) {
-  ENVOY_LOG(debug, "DNS Resolver flushing {} queries", queries_with_pending_cb_.size());
-  for (std::set<PendingResolution*>::iterator it = queries_with_pending_cb_.begin();
-       it != queries_with_pending_cb_.end(); ++it) {
-    auto query = *it;
-
-    ASSERT(query->pending_cb_);
-    query->callback_(query->pending_cb_->status_, std::move(query->pending_cb_->responses_));
-
-    if (query->owned_) {
-      ENVOY_LOG(debug, "Resolution for {} completed (async)", query->dns_name_);
-      delete *it;
-    } else {
-      ENVOY_LOG(debug, "Resolution for {} completed (synchronously)", query->dns_name_);
-      query->synchronously_completed_ = true;
-    }
-  }
-
-  // Purge the contents so no one tries to delete them again.
-  queries_with_pending_cb_.clear();
-
-  if (with_error) {
-    // The main sd ref is destroyed here because a callback with an error is good indication that
-    // the connection to the DNS server is faulty and needs to be torn down.
-    //
-    // Deallocation of the MainSdRef __has__ to happen __after__ flushing queries. Flushing queries
-    // de-allocates individual refs, so deallocating the main ref ahead would cause deallocation of
-    // invalid individual refs per dns_sd.h
-    deallocateMainSdRef();
-    initializeMainSdRef();
-  }
-}
+AppleDnsResolverImpl::PendingResolution::PendingResolution(AppleDnsResolverImpl& parent,
+                                                           ResolveCb callback,
+                                                           Event::Dispatcher& dispatcher,
+                                                           const std::string& dns_name)
+    : parent_(parent), callback_(callback), dispatcher_(dispatcher), dns_name_(dns_name),
+      pending_cb_({ResolutionStatus::Success, {}}) {}
 
 AppleDnsResolverImpl::PendingResolution::~PendingResolution() {
   ENVOY_LOG(debug, "Destroying PendingResolution for {}", dns_name_);
+
+  // dns_sd.h says:
+  //   If the reference's underlying socket is used in a run loop or select() call, it should
+  //   be removed BEFORE DNSServiceRefDeallocate() is called, as this function closes the
+  //   reference's socket.
+  sd_ref_event_.reset();
```

Comment:
what happens here if sd_ref_event_ was never initialized? (because fd was -1, etc) would this no-op or crash?

Reply:
it would no-op. `std::unique_ptr::reset()` destroys the currently managed object if any.